File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/observer/ObserverPattern.java
package jv.com.tencent.cain.design_pattern.observer;

/**
 * @author : jiangyu
 * @date : 2020/3/3
 * @desc : 观察者设计模式
 */
public class ObserverPattern {
    public static void main(String[] args) {
        Subject subject = new ConcreteSubject();

        Observer nbaObserver = new NBAObserver();
        Observer cartoonObserver = new CartoonObserver();

        subject.addObserver(nbaObserver);
        subject.addObserver(cartoonObserver);

        subject.notifyAllObserver("老师来了");
        subject.notify(nbaObserver, "老板进来了");
    }
}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/observer/ObserverPattern.kt
package kt.com.tencent.cain.desigin_pattern.observer

/**
 * @author : jiangyu
 * @date   : 2020/3/3
 * @desc   : 观察者设计模式
 */
fun main() {
    val subject = ConcreteSubject()

    val nbaObserver: Observer = NBAObserver()
    val cartoonObserver: Observer = CartoonObserver()

    subject.addObserver(nbaObserver)
    subject.addObserver(cartoonObserver)

    subject.notifyAll("老大回来了")
    subject.notify(cartoonObserver, "老师来了")
}